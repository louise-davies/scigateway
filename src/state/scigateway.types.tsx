import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadedAuthType = 'scigateway:loaded_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadScheduledMaintenanceStateType =
  'scigateway:load_scheduled_maintenance_state';
export const LoadMaintenanceStateType = 'scigateway:load_maintenance_state';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SendThemeOptionsType = 'scigateway:api:send_themeoptions';
export const LoadDarkModePreferenceType =
  'scigateway:load_dark_mode_preference';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';
export const RegisterStartUrlType = 'scigateway:register_start_url';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
  showHelpPageButton: boolean;
}

export interface StartUrlPayload {
  startUrl: string;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface SendThemeOptionsPayload {
  theme: Theme;
  broadcast: boolean;
}

export interface LoadDarkModePreferencePayload {
  darkMode: boolean;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface ScheduledMaintenanceStatePayLoad {
  scheduledMaintenance: ScheduledMaintenanceState;
}

export interface ScheduledMaintenanceState {
  show: boolean;
  message: string;
}

export interface MaintenanceStatePayLoad {
  maintenance: MaintenanceState;
}

export interface MaintenanceState {
  show: boolean;
  message: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
