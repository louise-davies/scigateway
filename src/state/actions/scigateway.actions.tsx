import axios from 'axios';
import { Action, AnyAction } from 'redux';
import log from 'loglevel';
import {
  ConfigureStringsType,
  ConfigureStringsPayload,
  ToggleDrawerType,
  AuthFailureType,
  AuthSuccessType,
  ApplicationStrings,
  SignOutType,
  FeatureSwitchesPayload,
  ConfigureFeatureSwitchesType,
  FeatureSwitches,
  LoadingAuthType,
  LoadedAuthType,
  DismissNotificationType,
  DismissNotificationPayload,
  RequestPluginRerenderType,
  AuthProviderPayload,
  LoadAuthProviderType,
  SiteLoadingType,
  SiteLoadingPayload,
  ConfigureAnalyticsPayload,
  ConfigureAnalyticsType,
  InitialiseAnalyticsType,
  ToggleHelpType,
  AddHelpTourStepsType,
  AddHelpTourStepsPayload,
  InvalidateTokenType,
  SendThemeOptionsType,
  SendThemeOptionsPayload,
  LoadDarkModePreferenceType,
  LoadDarkModePreferencePayload,
  StartUrlPayload,
  RegisterStartUrlType,
  ScheduledMaintenanceState,
  LoadScheduledMaintenanceStateType,
  ScheduledMaintenanceStatePayLoad,
  MaintenanceStatePayLoad,
  LoadMaintenanceStateType,
  MaintenanceState,
  NotificationType,
} from '../scigateway.types';
import { ActionType, ThunkResult, StateType } from '../state.types';
import loadMicroFrontends from './loadMicroFrontends';
import { push } from 'connected-react-router';
import { ThunkAction } from 'redux-thunk';
import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const configureStrings = (
  appStrings: ApplicationStrings
): ActionType<ConfigureStringsPayload> => ({
  type: ConfigureStringsType,
  payload: {
    res: appStrings,
  },
});

export const loadStrings = (path: string): ThunkResult<Promise<void>> => {
  return async (dispatch) => {
    await axios
      .get(path)
      .then((res) => {
        dispatch(configureStrings(res.data));
      })
      .catch((error) =>
        log.error(`Failed to read strings from ${path}: ${error}`)
      );
  };
};

export const loadFeatureSwitches = (
  featureSwitches: FeatureSwitches
): ActionType<FeatureSwitchesPayload> => ({
  type: ConfigureFeatureSwitchesType,
  payload: {
    switches: featureSwitches,
  },
});

export const registerStartUrl = (
  startUrl: string
): ActionType<StartUrlPayload> => ({
  type: RegisterStartUrlType,
  payload: {
    startUrl: startUrl,
  },
});

export const addHelpTourSteps = (
  steps: Step[]
): ActionType<AddHelpTourStepsPayload> => ({
  type: AddHelpTourStepsType,
  payload: {
    steps,
  },
});

export const loadAuthProvider = (
  authProvider: string,
  authUrl?: string
): ActionType<AuthProviderPayload> => ({
  type: LoadAuthProviderType,
  payload: {
    authProvider,
    authUrl,
  },
});

export const loadScheduledMaintenanceState = (
  scheduledMaintenance: ScheduledMaintenanceState
): ActionType<ScheduledMaintenanceStatePayLoad> => ({
  type: LoadScheduledMaintenanceStateType,
  payload: {
    scheduledMaintenance: scheduledMaintenance,
  },
});

export const loadMaintenanceState = (
  maintenance: MaintenanceState
): ActionType<MaintenanceStatePayLoad> => ({
  type: LoadMaintenanceStateType,
  payload: {
    maintenance: maintenance,
  },
});

export const loadingAuthentication = (): Action => ({
  type: LoadingAuthType,
});

export const loadedAuthentication = (): Action => ({
  type: LoadedAuthType,
});

export const unauthorised = (): Action => ({
  type: AuthFailureType,
});

export const authorised = (): Action => ({
  type: AuthSuccessType,
});

export const invalidToken = (): Action => ({
  type: InvalidateTokenType,
});

export const siteLoadingUpdate = (
  loading: boolean
): ActionType<SiteLoadingPayload> => ({
  type: SiteLoadingType,
  payload: {
    loading,
  },
});

export const configureAnalytics = (
  id: string
): ActionType<ConfigureAnalyticsPayload> => ({
  type: ConfigureAnalyticsType,
  payload: {
    id,
  },
});

export const initialiseAnalytics = (): Action => ({
  type: InitialiseAnalyticsType,
});

export const configureSite = (): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    await axios
      .get(`/settings.json`)
      .then((res) => {
        const settings = res.data;

        // invalid settings.json - Use JSON.parse to give detailed error info
        if (typeof settings !== 'object') {
          throw Error('Invalid format');
        }

        if (settings['ga-tracking-id']) {
          dispatch(configureAnalytics(settings['ga-tracking-id']));
        }

        dispatch(
          loadAuthProvider(settings['auth-provider'], settings['authUrl'])
        );

        const loadingPromises = [];

        const provider = getState().scigateway.authorisation.provider;

        if (provider.fetchMaintenanceState) {
          provider.fetchMaintenanceState().then((maintenanceState) => {
            dispatch(loadMaintenanceState(maintenanceState));
          });
        }

        // after auth provider is set then the token needs to be verified
        // also attempt to auto login if the auth provider allows it
        if (provider.isLoggedIn()) {
          const verifyingLogin = provider
            .verifyLogIn()
            .then(() => {
              dispatch(authorised());
            })
            .catch(() => {
              if (provider.autoLogin) {
                dispatch(loadingAuthentication());
                loadingPromises.push(
                  provider
                    .autoLogin()
                    .then(() => {
                      dispatch(authorised());
                    })
                    .catch(() => {
                      dispatch(invalidToken());
                    })
                );
              } else {
                dispatch(invalidToken());
              }
            });

          loadingPromises.push(verifyingLogin);
        } else if (provider.autoLogin) {
          dispatch(loadingAuthentication());
          loadingPromises.push(
            provider
              .autoLogin()
              .then(() => {
                dispatch(authorised());
              })
              .catch(() => {
                dispatch(loadedAuthentication());
              })
          );
        }

        if (settings['features']) {
          dispatch(loadFeatureSwitches(settings['features']));
        }

        dispatch(addHelpTourSteps(settings['help-tour-steps']));

        if (settings['startUrl']) {
          dispatch(registerStartUrl(settings['startUrl']));
        }

        const uiStringResourcesPath = !settings['ui-strings'].startsWith('/')
          ? '/' + settings['ui-strings']
          : settings['ui-strings'];
        const loadingResources = dispatch(loadStrings(uiStringResourcesPath));
        loadingPromises.push(loadingResources);

        const loadingPlugins = loadMicroFrontends.init(settings.plugins);
        loadingPromises.push(loadingPlugins);

        return Promise.all(loadingPromises).then(() => {
          dispatch(siteLoadingUpdate(false));
        });
      })
      .catch((error) => {
        log.error(`Error loading settings.json: ${error.message}`);
      });

    // load dark mode preference from local storage into store
    // otherwise, fetch system preference
    const darkModeLocalStorage = localStorage.getItem('darkMode');
    if (darkModeLocalStorage) {
      const preference = darkModeLocalStorage === 'true' ? true : false;
      dispatch(loadDarkModePreference(preference));
    } else {
      const mq = window.matchMedia('(prefers-color-scheme: dark)');
      if (mq) dispatch(loadDarkModePreference(mq.matches));
    }

    const provider = getState().scigateway.authorisation.provider;
    if (provider.fetchScheduledMaintenanceState) {
      provider
        .fetchScheduledMaintenanceState()
        .then((scheduledMaintenanceState) => {
          dispatch(loadScheduledMaintenanceState(scheduledMaintenanceState));

          // Checking the state in the GET response because it does not get
          // loaded into the store before this check is performed
          if (scheduledMaintenanceState['show']) {
            displayScheduledMaintenanceBanner(
              scheduledMaintenanceState['message']
            );
          }
        });
    }
  };
};

const displayScheduledMaintenanceBanner = (message: string): void => {
  document.dispatchEvent(
    new CustomEvent('scigateway', {
      detail: {
        type: NotificationType,
        payload: {
          severity: 'warning',
          message: message,
        },
      },
    })
  );
};

export const toggleDrawer = (): Action => ({
  type: ToggleDrawerType,
});

export const toggleHelp = (): Action => ({
  type: ToggleHelpType,
});

export const signOut = (): ThunkAction<void, StateType, null, AnyAction> => (
  dispatch
) => {
  dispatch({ type: SignOutType });
  dispatch(push('/'));
};

export const verifyUsernameAndPassword = (
  username: string,
  password: string
): ThunkResult<Promise<void>> => {
  return async (dispatch, getState) => {
    // will be replaced with call to login API for authentification
    dispatch(loadingAuthentication());
    const authProvider = getState().scigateway.authorisation.provider;
    await authProvider
      .logIn(username, password)
      .then(() => {
        dispatch(authorised());

        // redirect the user to the original page they were trying to get to
        // the referrer is added by the redirect in routing.component.tsx
        const previousRouteState = getState().router.location.state;
        dispatch(
          push(
            previousRouteState && previousRouteState.referrer
              ? previousRouteState.referrer
              : '/'
          )
        );
      })
      .catch(() => {
        // probably want to do something smarter with
        // err.response.status (e.g. 403 or 500)
        dispatch(unauthorised());
      });
  };
};

export const requestPluginRerender = (): ActionType<{
  broadcast: boolean;
}> => ({
  type: RequestPluginRerenderType,
  payload: {
    broadcast: true,
  },
});

export const sendThemeOptions = (
  theme: Theme
): ActionType<SendThemeOptionsPayload> => ({
  type: SendThemeOptionsType,
  payload: {
    theme,
    broadcast: true,
  },
});

export const loadDarkModePreference = (
  darkMode: boolean
): ActionType<LoadDarkModePreferencePayload> => ({
  type: LoadDarkModePreferenceType,
  payload: {
    darkMode: darkMode,
  },
});

export const dismissMenuItem = (
  index: number
): ActionType<DismissNotificationPayload> => {
  return {
    type: DismissNotificationType,
    payload: {
      index,
    },
  };
};
