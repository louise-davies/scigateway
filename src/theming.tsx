import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#FF6900',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  primary: string;
  secondary: string;
  background: string;
  paper: string;
  blue: string;
  orange: string;
  red: string;
  grey: string;
}

const DARK_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

//For experimenting
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const DARK_MODE_COLOURS_HIGH_CONTRAST: ThemeColours = {
  primary: '#86B4FF',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#003088',
  background: '#FAFAFA',
  paper: '#FFF',
  blue: '#003088',
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
};

export const buildTheme = (darkModePreference: boolean): Theme => {
  let options: UKRIThemeOptions;
  const colours = darkModePreference ? DARK_MODE_COLOURS : LIGHT_MODE_COLOURS;

  const overrides = {
    MuiLink: {
      root: {
        color: colours.blue,
      },
    },
    MuiTabs: {
      indicator: {
        color: STATIC_COLOURS.darkBlue,
        textDecoration: 'underline',
      },
    },
    MuiFormLabel: {
      root: {
        '&$error': {
          color: colours.red,
        },
        '&$focused': {
          color: colours.blue,
        },
      },
      asterisk: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiBadge: {
      colorPrimary: {
        backgroundColor: STATIC_COLOURS.orange,
      },
    },
    MuiInput: {
      underline: {
        '&$error:after': {
          borderBottomColor: colours.red,
        },
        '&:after': {
          borderBottomColor: colours.blue,
        },
      },
    },
    MuiOutlinedInput: {
      root: {
        '&$error $notchedOutline': {
          borderColor: colours.red,
        },
      },
    },
    MuiFormHelperText: {
      root: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiPickersToolbar: {
      toolbar: {
        backgroundColor: STATIC_COLOURS.darkBlue,
      },
    },
    MuiPickersCalendarHeader: {
      dayLabel: {
        color: colours.grey,
      },
    },
    MuiPickersDay: {
      current: {
        color: colours.blue,
      },
      dayDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersYear: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      yearSelected: {
        color: colours.blue,
      },
      yearDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersMonth: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      monthSelected: {
        color: colours.blue,
      },
      monthDisabled: {
        color: colours.grey,
      },
    },
  };

  if (darkModePreference) {
    options = {
      palette: {
        // Light/dark mode
        type: 'dark',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#257fff',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  } else {
    options = {
      palette: {
        // Light/dark mode
        type: 'light',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#1E5DF8',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  }

  return createMuiTheme(options);
};

function mapThemeProviderStateToProps(
  state: StateType
): { prefersDarkMode: boolean } {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference: boolean = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  return (
    <MuiThemeProvider theme={buildTheme(darkModePreference)}>
      {props.children}
    </MuiThemeProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
