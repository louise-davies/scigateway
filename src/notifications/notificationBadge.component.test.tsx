import React from 'react';
import NotificationBadge, {
  UnconnectedNotificationBadge,
  CombinedNotificationBadgeProps,
} from './notificationBadge.component';
import Badge from '@mui/material/Badge';
import { ThemeProvider, StyledEngineProvider } from '@mui/material/styles';
import { buildTheme } from '../theming';
import configureStore from 'redux-mock-store';
import { authState, initialState } from '../state/reducers/scigateway.reducer';
import { dismissMenuItem } from '../state/actions/scigateway.actions';
import { Provider } from 'react-redux';
import { StateType } from '../state/state.types';
import { mount, shallow } from 'enzyme';

describe('Notification Badge component', () => {
  const theme = buildTheme(false);

  let mockStore;
  let state: StateType;
  let props: CombinedNotificationBadgeProps;

  beforeEach(() => {
    mockStore = configureStore();

    state = {
      scigateway: { ...initialState, authorisation: { ...authState } },
    };
    props = {
      notifications: [{ message: 'my message', severity: 'warning' }],
      deleteMenuItem: jest.fn(),
    };
  });

  it('Notification badge renders correctly', () => {
    const wrapper = shallow(<UnconnectedNotificationBadge {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correct number of notifications in the badge', () => {
    let wrapper = shallow(
      <UnconnectedNotificationBadge
        {...props}
        notifications={[
          ...props.notifications,
          { message: 'my other message', severity: 'success' },
        ]}
      />
    );

    expect(wrapper.find(Badge).prop('badgeContent')).toEqual(2);

    wrapper = shallow(
      <UnconnectedNotificationBadge
        {...props}
        notifications={[]}
        deleteMenuItem={props.deleteMenuItem}
      />
    );

    expect(wrapper.find(Badge).prop('badgeContent')).toBeNull();
  });

  it('sends dismissMenuItem action when dismissNotification prop is called', () => {
    state.scigateway.notifications = props.notifications;
    const testStore = mockStore(state);

    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <NotificationBadge />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper
      .find('[aria-label="Open notification menu"]')
      .last()
      .simulate('click');
    wrapper.update();

    wrapper
      .find('[aria-label="Dismiss notification"]')
      .last()
      .simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(dismissMenuItem(0));
  });

  it('opens menu when button clicked and closes menu when there are no more notifications', () => {
    state.scigateway.notifications = props.notifications;

    let testStore = mockStore(state);

    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <NotificationBadge />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper
      .find('[aria-label="Open notification menu"]')
      .last()
      .simulate('click');

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();

    state.scigateway.notifications = [];
    testStore = mockStore(state);
    wrapper.setProps({ store: testStore });

    expect(wrapper.find('#notifications-menu').exists()).toBeFalsy();
  });

  it('opens menu with no notifications message when button clicked and there are no notifications', () => {
    state.scigateway.notifications = [];

    const testStore = mockStore(state);

    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <NotificationBadge />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper
      .find('[aria-label="Open notification menu"]')
      .last()
      .simulate('click');

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeTruthy();
  });

  it('no notifications message disappears when a notification occurs', () => {
    state.scigateway.notifications = [];

    let testStore = mockStore(state);

    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <NotificationBadge />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper
      .find('[aria-label="Open notification menu"]')
      .last()
      .simulate('click');

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeTruthy();

    state.scigateway.notifications = props.notifications;
    testStore = mockStore(state);
    wrapper.setProps({ store: testStore });

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeFalsy();
  });

  it('no notifications message can be closed', () => {
    state.scigateway.notifications = [];
    const testStore = mockStore(state);

    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <NotificationBadge />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    //Check can close using close button
    wrapper
      .find('[aria-label="Open notification menu"]')
      .last()
      .simulate('click');

    expect(wrapper.find('#notifications-menu').exists()).toBeTruthy();
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeTruthy();

    wrapper
      .find('[aria-label="Dismiss notification"]')
      .last()
      .simulate('click');

    expect(wrapper.find('#notifications-menu').exists()).toBeFalsy();
    expect(
      wrapper.find('[aria-label="No notifications message"]').exists()
    ).toBeFalsy();
  });
});
