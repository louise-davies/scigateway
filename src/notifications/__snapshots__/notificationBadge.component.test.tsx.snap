// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification Badge component Notification badge renders correctly 1`] = `
<div
  className="tour-notifications"
>
  <ForwardRef(IconButton)
    aria-haspopup="true"
    aria-label="Open notification menu"
    onClick={[Function]}
    size="large"
    sx={
      Object {
        "color": "primary.contrastText",
        "margin": 1,
      }
    }
  >
    <ForwardRef(Badge)
      badgeContent={1}
      color="primary"
    >
      <Memo(ForwardRef(NotificationsIcon)) />
    </ForwardRef(Badge)>
  </ForwardRef(IconButton)>
  <ForwardRef(Menu)
    anchorEl={null}
    id="notifications-menu"
    onClose={[Function]}
    open={false}
  >
    <ForwardRef(ScigatewayNotification)
      dismissNotification={[Function]}
      index={0}
      key="0"
      message="my message"
      severity="warning"
    />
  </ForwardRef(Menu)>
</div>
`;
